### ISSUE TITLE
Build fails when using supabase-js

### ISSUE THREAD

**Issue by aiji42** [remix-run/remix#2075]
### What version of Remix are you using?

1.2.2

### Steps to Reproduce

1. Select Cloudflare workers (or pages) as runtime.
2. Install `@supabase/supabase-js`. 
3. Call createClient in loader or .server.ts
4. Start up a dev server (`yarn dev`)

### Expected Behavior

The server comes up without error.

### Actual Behavior

Error occurs. `TypeError: globalThis.XMLHttpRequest is not a constructor`
![スクリーンショット 2022-02-22 午後8 21 26](https://user-images.githubusercontent.com/6711766/155122543-5754d1fe-5797-454b-8a1d-39e75a9710a6.png)

**Comment by aiji42** [comment]
You can work around it by adapting such a patch.
If the runtime is Cloudflare Workers, the browser module should be the preferred solution.

```
diff --git a/node_modules/@remix-run/dev/compiler.js b/node_modules/@remix-run/dev/compiler.js
index 0705b93..72ed487 100644
--- a/node_modules/user/node_modules/@remix-run/dev/compiler.js
+++ b/node_modules/user/node_modules/@remix-run/dev/compiler.js
@@ -364,7 +364,7 @@ async function createServerBuild(config, options, assetsManifestPromiseRef) {
     format: config.serverModuleFormat,
     treeShaking: true,
     minify: options.mode === build$1.BuildMode.Production && !!config.serverBuildTarget && ["cloudflare-workers", "cloudflare-pages"].includes(config.serverBuildTarget),
-    mainFields: config.serverModuleFormat === "esm" ? ["module", "main"] : ["main", "module"],
+    mainFields: ["browser", "module", "main"],
     target: options.target,
     inject: [reactShim],
     loader: loaders.loaders,
```

**Comment by and3rsonsousa** [comment]
It also happens with graphql-request.

**Comment by Pitasi** [comment]
Yeah it's happening to me too when using graphql-request. The small patch proposed by @aiji42 didn't help: it moved the problem from build-time to run-time (I get an application error in the browser: `ReferenceError: XMLHttpRequest is not defined`).

**Comment by Pitasi** [comment]
Yes, I'm using v1.2.2, and yes I updated createServerBuild. I think that graphql-request tries to determine if fetch is available or not in some way but it doesn't work.

Anyway see my prev comment that I edited, I got it to work.

Thanks for helping!

**Comment by simjes** [comment]
Same error goes for `@notionhq/client`

**Comment by kpalmvik** [comment]
I see the same error when trying to use [@sanity/client](https://www.sanity.io/docs/js-client) in a vanilla Remix 1.2.3 app with Cloudflare Pages as deploy target:

```
...
Compiled Worker successfully.
[pages:err] TypeError: globalThis.XMLHttpRequest is not a constructor
    at checkTypeSupport (/somepath/remix-starter/node_modules/rollup-plugin-node-polyfills/polyfills/http-lib/capability.js:20:11)
    at node_modules/rollup-plugin-node-polyfills/polyfills/http-lib/capability.js (/somepath/remix-starter/node_modules/rollup-plugin-node-polyfills/polyfills/http-lib/capability.js:39:45)
    at __init (/var/folders/f5/62nm5zbj5ss915gv70f465100000gn/T/functionsWorker.js:39:56)
    at node_modules/rollup-plugin-node-polyfills/polyfills/http-lib/request.js (/somepath/remix-starter/node_modules/rollup-plugin-node-polyfills/polyfills/http-lib/request.js:1:1)
    at __init (/var/folders/f5/62nm5zbj5ss915gv70f465100000gn/T/functionsWorker.js:39:56)
    at node-modules-polyfills:https (/somepath/remix-starter/functions/node-modules-polyfills:https:30:1)
    at __init (/var/folders/f5/62nm5zbj5ss915gv70f465100000gn/T/functionsWorker.js:39:56)
    at node-modules-polyfills-commonjs:https (/somepath/remix-starter/functions/node-modules-polyfills-commonjs:https:2:18)
    at __require (/var/folders/f5/62nm5zbj5ss915gv70f465100000gn/T/functionsWorker.js:42:50)
    at node_modules/eventsource/lib/eventsource.js (/somepath/remix-starter/node_modules/eventsource/lib/eventsource.js:4:13)
Could not start Miniflare.
ERROR: "dev:wrangler" exited with 1.

```

**Comment by aiji42** [comment]
Thanks everyone for the reports.
I have already created a PR to solve this problem already but it is not merged easily.
As I indicated at the beginning, patching is one way to deal with the problem, but it is complicated to have to fix the patch together with every version upgrade on the Remix side.

I have therefore developed a library that allows overriding the esbuild settings in Remix, and I would like to introduce it to you.
https://github.com/aiji42/remix-esbuild-override

```bash
yarn add -D remix-esbuild-override
```

Update `scripts.postinstall` in package.json. Then run `yarn instal` again to run `postinstall`.

```json
"scripts": {
  "postinstall": "yarn remix setup cloudflare-workers && yarn remix-esbuild-override"
}
```

Update `remix.config.js`

```js
// remix.config.js
/**
 * @type {import('remix-esbuild-override').AppConfig}
 */
module.exports = {
  serverBuildTarget: "cloudflare-workers",
  server: "./server.js",
  devServerBroadcastDelay: 1000,
  ignoredRouteFiles: [".*"],
  esbuildOverride: (option, { isServer }) => {
    if (isServer) option.mainFields = ["browser", "module", "main"];

    return option;
  },
};
```
It will allow us to override esbuild settings, which will not only address this issue, but also various other compiler-dependent issues.

---
I submitted a PR for this feature to be added to the Remix side, but it was rejected because it was conceptually unacceptable. Therefore, I am developing this remix-esbuild-override as a third-party library.
